### PR TITLE
dvdplayer: avoid unwanted dropping in output stage of video

### PR DESCRIPTION
**Pull request by FernetMenta**
fixes uncontrolled dropping of late video frames

### CODE REVIEW

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by ace20022** [comment]
What's the image of ```DVD_TIME_TO_MSEC(iSleepTime)```?

**Comment by FernetMenta** [comment]
can you rephrase the question? what do you mean by "image of ..."

**Comment by ace20022** [comment]
Just out of interest I wondered "how negative it can be". 
I had a look at ```CXBMCRenderManager::WaitForBuffer(volatile bool& bStop, int timeout)```
Now we have ```min(50, max(x, 50)) = 50 for all x```

Regarding image: https://en.wikipedia.org/wiki/Image_%28mathematics%29.

**Comment by FernetMenta** [comment]
iSleepTime can indeed be negative if a frame is late. The old code used 1ms as timeout in this case which caused the frame to be dropped very likely.
Assume timeout is 500. CXBMCRenderManager::WaitForBuffer would not wait longer than 50ms in a single cycle in order to better observe stop.

**Comment by fritsch** [comment]
That will increase the skip counter most likely - if we wait 50ms once - even a 24p movie will skip at least one time. Is that intended?

**Comment by FernetMenta** [comment]
not sure if you understood what is happening here. this change only affects frames which are already late with negative sleep time. the old code silently dropped those.

**Comment by fritsch** [comment]
queue: -40, -50, -60, -70, -80

Not sore if you got my question, but - what would happen with the above queue?

**Comment by FernetMenta** [comment]
dropping is controlled by CDVDPlayerVideo::CalcDropRequirement. It does observe the entire queue and acts accordingly. 
You have not given enough information to tell what's happening to that queue. where exactly are those frames? If those are already in the render buffers, renderer would skip all late frames. And this wold happen with or without this change.
Again: the max timeout has nothing to do with how the queue is processed.

**Comment by FernetMenta** [comment]
The timeout given to WaitForBuffer is for the purpose of fail-safe. We want to submit the frame to the render buffers regardless of its timestamp.

**Comment by fritsch** [comment]
After discussing that with @FernetMenta on IRC - I understand why this is done as is and I am fine with it. It resolves a really big design flaw from a time without buffering. Should go to Helix, too :-)
